File: PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> rets = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        DFS(root, sum, rets, path);
        
        return rets;
    }
    
    public void DFS(TreeNode root, int sum, List<List<Integer>> rets, List<Integer> path) {
        if (root == null)
            return;
        
        path.add(root.val);
        sum -= root.val;
        if (root.left == null && root.right == null && sum == 0)
            rets.add(new ArrayList<>(path));
        
        DFS(root.left, sum, rets, path);
        DFS(root.right, sum, rets, path);
        sum += root.val;
        path.remove(path.size() - 1);
    }
}

File: lulu/pathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.ans = []
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.pathSumHelper(root, sum, [])
        return self.ans
    def pathSumHelper(self, root, gap, cur):
        if root is None:
            return
        if root.left is None and root.right is None:
            if gap == root.val:
                cur.append(root.val)
                self.ans.append(cur)
                return
        
        cur.append(root.val)
        self.pathSumHelper(root.left, gap - root.val, cur[:])
        self.pathSumHelper(root.right, gap - root.val, cur[:])
